# Introduction

Engage Science is the first to use purely external data to decode the complex relationships between people, roles, companies, and dynamic markets to
- map employees diversity,
- their flow between companies,
- and forecast expensive talent turnover risk.

The science is based on three decades of validated academic and industry research of the events that lead to turnover. It relies on long term studies of career progression and the market forces impacting engagement, atrition and talent movement.


# Objective

- To understand the datasets such as the variable information, the data types etc.
- Explore the datasets in order to extract useful and rightful information out of them.
- Check for data inconsistency.
- Look for any missing values in the datasets.
- Visualize the data for better understanding.

Let's get started.

### Import
Import the essential libarires

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### Read the data
- Read the Engage Talent's 'Talent Retention Risk' Dataset into `engage` dataframe.
- Parse the `date` column as date for future use.

In [2]:
engage = pd.read_csv("engage_trr_history_20190327.csv", parse_dates=["date"])
engage.head()

Unnamed: 0          isin exchange symbol       date  score
0           1  AED000901010    Dubai    DFM 2015-07-01     39
1           2  AED000901010    Dubai    DFM 2015-07-02     39
2           3  AED000901010    Dubai    DFM 2015-07-03     37
3           4  AED000901010    Dubai    DFM 2015-07-04     37
4           5  AED000901010    Dubai    DFM 2015-07-05     37

Drop the unnecessary columns.

In [3]:
engage.drop(['Unnamed: 0'], axis = 1, inplace = True)
engage.head()

isin exchange symbol       date  score
0  AED000901010    Dubai    DFM 2015-07-01     39
1  AED000901010    Dubai    DFM 2015-07-02     39
2  AED000901010    Dubai    DFM 2015-07-03     37
3  AED000901010    Dubai    DFM 2015-07-04     37
4  AED000901010    Dubai    DFM 2015-07-05     37

### Understanding Data

- Total number of rows in the dataset: 2567187
- Total number of columns in the dataset: 5

In [4]:
engage.shape

(2567187, 5)

Here, we have 
- 3 columns (`isin, exchange, symbol`) as object data-type
- 1 column (`date`) as datetime data-type
- 1 column (`score`) as integer data-type

In [5]:
engage.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2567187 entries, 0 to 2567186
Data columns (total 5 columns):
isin        object
exchange    object
symbol      object
date        datetime64[ns]
score       int64
dtypes: datetime64[ns](1), int64(1), object(3)
memory usage: 97.9+ MB


Exploring the `score` field:
- total count = 2567187
- mean of values = 39.69822
- minimum value = 2
- maximum value = 92

In [6]:
engage.describe()

score
count  2.567187e+06
mean   3.969822e+01
std    9.504913e+00
min    2.000000e+00
25%    3.400000e+01
50%    3.900000e+01
75%    4.500000e+01
max    9.200000e+01

Exploring the `date` field:
- total count = 2567187
- number of unique values = 1366 means the total number of days captured.
- range of date is from 2015-07-01 to 2019-03-27

In [7]:
engage.describe(include = ["datetime64"])

date
count               2567187
unique                 1366
top     2018-01-09 00:00:00
freq                   1882
first   2015-07-01 00:00:00
last    2019-03-27 00:00:00

Exploring the rest of the fields:
- total count = 2567187 for all the fields. This assures that there is no missing value in the dataset.
- `isin` has 1977 unique values and `symbol` has 1856 unique values listed in 33 different `exchanges`

In [8]:
engage.describe(include = ['object'])

isin exchange   symbol
count        2567187  2567187  2567187
unique          1977       33     1856
top     US8960475031     NYSE       DG
freq            1366  1136777     2732


##### Since we have 1366 days data, grouping the dataset by `symbol` to find the data consistency across dataset.

In [104]:
# symbols 
engage.groupby(['symbol'])['date'].count().reset_index()


symbol  date
0     105560  1366
1     126560  1366
2       2002  1249
3       2317  1366
4       2332  1353
5       2353  1366
6       2354  1366
7       2395  1366
8       2454  1366
9       2498  1366
10      2610  1366
11      2802  1366
12      3401  1366
13      3402  1366
14      3407  1366
15      3659  1366
16      3771  1345
17      3842  1249
18      4005  1366
19      4188  1366
20      4280  1281
21      4324  1366
22      4502  1366
23      4503  1366
24      4523  1366
25      4568  1366
26      4704  1366
27      4755  1366
28      4901  1366
29      4902  1366
...      ...   ...
1826    WYNN  1366
1827       X  2730
1828     XEC  1366
1829    XENT  1366
1830    XLNX  1366
1831    XLRN  1366
1832     XOM  1366
1833    XOMA  1366
1834     XON  1261
1835     XPO  1366
1836    XRAY  1366
1837     XRX  1366
1838     XYL  1366
1839     YAR  1366
1840    YELP  1366
1841     YRI  1366
1842     YUM  1366
1843       Z  1366
1844     Z74  1366
1845    ZAGG  1366
1846    ZAYO  1366
1847     ZBH  1366
1848    ZBRA  1366
1849     ZEN  1366
1850    ZGNX  1366
1851    ZIOP  1366
1852    ZNGA  1366
1853     ZTS  1366
1854    ZUMZ  1366
1855    ZURN  1330

[1856 rows x 2 columns]

In [10]:
engage.groupby(['symbol'])['date'].count().describe()

count    1856.000000
mean     1383.182651
std       161.060202
min      1249.000000
25%      1366.000000
50%      1366.000000
75%      1366.000000
max      2732.000000
Name: date, dtype: float64

##### 2 out of 5 symbols above don't show data consistency. Exploring the values further.
- Max value = 2732 and Min Value = 1249 with standard deviation of 161.

 ******************************************************************************************************************

In [ ]:
# By Evan

There are two different forms of ‘symbol’:
- 1765 unique non-numerical symbols.
- 91 unique numerical symbols.
- 1856 unique symbols in total.

In [19]:
non_num_symbol=[x for x in engage['symbol'] if not x.isdigit()]
len(set(non_num_symbol))

1765

In [20]:
num_symbol=[n for n in engage['symbol'] if n.isdigit()]
len(set(num_symbol))

91

In [21]:
# also check the forms of 'exchange'
non_num_symbol=[x for x in engage['exchange'] if not x.isdigit()]
len(set(non_num_symbol))

33

### Columns combination

In [53]:
engage_new = engage.copy()

In [107]:
df = engage.copy()
df["new_symbol"] = df["exchange"] + "_" + df["symbol"]

In [54]:
engage_new['ticker'] = engage_new['symbol'].str.cat(engage_new['exchange'],sep="_")

engage_new = engage_new[['isin','ticker','date','score']]

engage_new.head()

isin     ticker       date  score
0  AED000901010  DFM_Dubai 2015-07-01     39
1  AED000901010  DFM_Dubai 2015-07-02     39
2  AED000901010  DFM_Dubai 2015-07-03     37
3  AED000901010  DFM_Dubai 2015-07-04     37
4  AED000901010  DFM_Dubai 2015-07-05     37

### Duplicates Checkup

In [56]:
#Show all the tickers that have more than one isin_id
tickers = engage_new.groupby("ticker").nunique()["isin"].reset_index().query("isin != 1").sort_values("isin", ascending=False)
tickers

ticker  isin
98                 AA_NYSE     4
917                IHG_LSE     4
769               FTI_NYSE     4
308              BIOC_NASD     3
1110              MDR_NYSE     3
866                HK_NYSE     3
401               CCE_NYSE     3
356               BTU_NYSE     3
1861              XRX_NYSE     3
221              ARNA_NASD     2
1431              REN_NYSE     2
1345             PFSI_NYSE     2
1233               NM_NYSE     2
1229             NLSN_NYSE     2
1187              MTW_NYSE     2
1144             MNKD_NASD     2
113              ACLS_NASD     2
1080             MACK_NASD     2
1054             LOGI_NASD     2
1041              LLL_NYSE     2
1007              KOS_NYSE     2
974               JMAT_LSE     2
966               JCI_NYSE     2
957               ITT_NYSE     2
118              ACXM_NASD     2
908              ICON_NASD     2
1414              QTM_NYSE     2
1496             SBAC_NASD     2
222              ARRS_NASD     2
1564              SPB_NYSE     2
...                    ...   ...
392               CBB_NYSE     2
385              CATM_NASD     2
363               BXS_NYSE     2
195                AM_NYSE     2
335              BOFI_NASD     2
202              ANTM_NYSE     2
299                BEZ_LSE     2
287             BB_Toronto     2
274               BAS_NYSE     2
231               ASH_NYSE     2
224              ARWR_NASD     2
446                CI_NYSE     2
448               CLF_NYSE     2
477               CNX_NYSE     2
760              FOXA_NASD     2
815             GOOGL_NASD     2
795          GIB.A_Toronto     2
101   ABI_EuronextBrussels     2
785               GDI_NYSE     2
771               FTR_NASD     2
152              AINV_NASD     2
751               FNF_NYSE     2
497                CPG_LSE     2
714              FCEL_NASD     2
641            ELD_Toronto     2
640              EKSO_NASD     2
592                DK_NYSE     2
511               CRDA_LSE     2
509               CRC_NYSE     2
788               GEO_NYSE     2

[83 rows x 2 columns]

In [111]:
#count those non-unique isin_id
a = tickers.ticker.values
duplicate_df = engage_new[engage_new.ticker.isin(a)]
b = duplicate_df.groupby('ticker')['isin'].value_counts()
b

ticker                  isin        
AA_NYSE                 US0138175072     875
                        US0138721065     300
                        US0138171014     153
                        US1387210636      38
ABI_EuronextBrussels    BE0974293251     980
                        BE0003793107     386
ACLS_NASD               US0545402085    1024
                        US0545401095     342
ACXM_NASD               US0051251090    1190
                        US53815P1084     132
AGN_NYSE                USG0177J1081    1249
                        USIE00BY9D86     117
AINV_NASD               US03761U1060    1254
                        US03761U5020     112
AKZA_EuronextAmsterdam  NL0000009132    1281
                        NL0013267909      85
AM_NYSE                 US03673L1035    1353
                        US03676B1026      13
ANTM_NYSE               US94973V1070    1249
                        US0367521038     117
ARNA_NASD               US0400476075     736
                  

******************************************************************************************************************

## Date Gap Checkup

In [118]:
df1 = engage_new.groupby('ticker')['date'].count().reset_index()
df1.columns=['ticker','Date_count']
df1.head()

ticker  Date_count
0  105560_Korea        1366
1  126560_Korea        1366
2   2002_Taiwan        1249
3   2317_Taiwan        1366
4   2332_Taiwan        1353

In [119]:
df2=df1[df1['Date_count']<1366]
df2

ticker  Date_count
2           2002_Taiwan        1249
4           2332_Taiwan        1353
16           3771_Tokyo        1345
17           3842_Tokyo        1249
20     4280_SaudiArabia        1281
40           6664_Tokyo        1339
41           6676_Tokyo        1349
74           8247_Tokyo        1249
84           8795_Tokyo        1323
118           ACXM_NASD        1322
138            AET_NYSE        1338
145          AGM.A_NYSE        1358
203            ANW_NYSE        1315
218   ARG_EuronextParis        1356
293            BD15_LSE        1313
299             BEZ_LSE        1249
321             BLT_LSE        1330
335           BOFI_NASD        1279
352          BSC_Manila        1338
389             CA_NASD        1325
394           CBOE_NASD        1259
401            CCE_NYSE        1303
427   CGG_EuronextParis        1259
451            CLLN_LSE        1318
474         CNQ_Toronto        1364
478           COBZ_NASD        1283
485            COL_NYSE        1336
508           CRCM_NYSE        1364
547            CVG_NYSE        1283
609            DSW_NYSE        1365
...                 ...         ...
1268           NWL_NYSE        1348
1275      OANDO_Nigeria        1353
1278          OCLR_NASD        1351
1337          PERY_NASD        1308
1347            PF_NYSE        1304
1353           PHH_NYSE        1282
1412          QSII_NASD        1255
1429           REG_NYSE        1308
1473          RSYS_NASD        1353
1485      S51_Singapore        1361
1510            SDR_LSE        1360
1529          SHLD_NASD        1335
1530          SHLM_NASD        1249
1543            SKY_LSE        1304
1560          SONC_NASD        1356
1576         SREN_Swiss        1285
1607           SVU_NYSE        1301
1613           SXC_NYSE        1364
1621          SYNT_NASD        1293
1643           TER_NYSE        1329
1685        TRQ_Toronto        1330
1742          UNFI_NASD        1359
1765            VED_LSE        1273
1775           VLY_NYSE        1288
1816   WHL_Johannesburg        1297
1832      WOW_Australia        1360
1846           WTW_NYSE        1280
1858           XON_NYSE        1261
1864          X_Toronto        1364
1880         ZURN_Swiss        1330

[79 rows x 2 columns]

In [121]:
ticker_check_list = df2['ticker'].values

array(['2002_Taiwan', '2332_Taiwan', '3771_Tokyo', '3842_Tokyo',
       '4280_SaudiArabia', '6664_Tokyo', '6676_Tokyo', '8247_Tokyo',
       '8795_Tokyo', 'ACXM_NASD', 'AET_NYSE', 'AGM.A_NYSE', 'ANW_NYSE',
       'ARG_EuronextParis', 'BD15_LSE', 'BEZ_LSE', 'BLT_LSE', 'BOFI_NASD',
       'BSC_Manila', 'CA_NASD', 'CBOE_NASD', 'CCE_NYSE',
       'CGG_EuronextParis', 'CLLN_LSE', 'CNQ_Toronto', 'COBZ_NASD',
       'COL_NYSE', 'CRCM_NYSE', 'CVG_NYSE', 'DSW_NYSE', 'EDR_NYSE',
       'EGN_NYSE', 'ELD_Toronto', 'FCE.A_NYSE', 'FDUS_NASD', 'FSIC_NYSE',
       'FXJ_Australia', 'GIL_Toronto', 'IBKR_NASD', 'INF_LSE', 'INVP_LSE',
       'JSE_Johannesburg', 'KERX_NASD', 'KLXI_NASD', 'KMG_NYSE',
       'KS_NYSE', 'LPNT_NASD', 'MDXG_NASD', 'MNW_Toronto', 'NWL_NYSE',
       'OANDO_Nigeria', 'OCLR_NASD', 'PERY_NASD', 'PF_NYSE', 'PHH_NYSE',
       'QSII_NASD', 'REG_NYSE', 'RSYS_NASD', 'S51_Singapore', 'SDR_LSE',
       'SHLD_NASD', 'SHLM_NASD', 'SKY_LSE', 'SONC_NASD', 'SREN_Swiss',
       'SVU_NYSE', 'SXC_

In [207]:
engage_new.head()

isin     ticker       date  score
0  AED000901010  DFM_Dubai 2015-07-01     39
1  AED000901010  DFM_Dubai 2015-07-02     39
2  AED000901010  DFM_Dubai 2015-07-03     37
3  AED000901010  DFM_Dubai 2015-07-04     37
4  AED000901010  DFM_Dubai 2015-07-05     37

In [217]:
engage_new = engage_new.sort_values(["ticker", "date"])
engage_new["next_date"] = engage_new.groupby("ticker")["date"].shift(-1)
engage_new["date_gap"] = engage_new.apply(lambda x: (x["next_date"] - x["date"]).days, axis=1)

In [219]:
engage_new.head()

isin        ticker       date  score  next_date  date_gap
395294  KR7105560007  105560_Korea 2015-07-01     40 2015-07-02       1.0
395295  KR7105560007  105560_Korea 2015-07-02     40 2015-07-03       1.0
395296  KR7105560007  105560_Korea 2015-07-03     40 2015-07-04       1.0
395297  KR7105560007  105560_Korea 2015-07-04     41 2015-07-05       1.0
395298  KR7105560007  105560_Korea 2015-07-05     41 2015-07-06       1.0

In [222]:
engage_new[(engage_new["date_gap"].notnull()) & (engage_new["date_gap"] != 1)].sort_values("date_gap", ascending=False)

isin            ticker       date  score  next_date  date_gap
119055   CH0126881561        SREN_Swiss 2018-11-30     42 2019-02-20      82.0
2564449  ZAE000063863  WHL_Johannesburg 2018-11-30     42 2019-02-08      70.0
215749   GB0008754466          BD15_LSE 2019-01-12     37 2019-03-07      54.0
213022   GB0007365546          CLLN_LSE 2018-11-30     42 2019-01-18      49.0
2565756  ZAE000079711  JSE_Johannesburg 2018-12-10     42 2019-01-23      44.0
102448   CA9004351081       TRQ_Toronto 2019-01-30     39 2019-03-08      37.0
109064   CH0011075394        ZURN_Swiss 2018-11-30     42 2019-01-06      37.0
418446   PHY0718K1017        BSC_Manila 2019-01-30     39 2019-02-28      29.0
249799   GB00B17BBQ50          INVP_LSE 2019-02-27     41 2019-03-19      20.0
308094   JP3197740008        6664_Tokyo 2018-11-30     42 2018-12-19      19.0
404815   NGOANDO00002     OANDO_Nigeria 2019-02-15     53 2019-03-01      14.0
1155548  US3131481083        AGM.A_NYSE 2019-02-06     39 2019-02-15       9.0
2311703  US9111631035         UNFI_NASD 2019-03-19     39 2019-03-27       8.0
62900    CA2849025093       ELD_Toronto 2019-01-24     40 2019-02-01       8.0
27185    AU000000WOW2     WOW_Australia 2018-11-30     42 2018-12-07       7.0
197473   GB0002405495           SDR_LSE 2019-01-22     41 2019-01-29       7.0
426545   SG1H97877952     S51_Singapore 2019-02-15     55 2019-02-21       6.0
72461    CA3759161035       GIL_Toronto 2019-01-30     39 2019-02-05       6.0
269975   GB00BMJ6DW54           INF_LSE 2019-01-29     40 2019-02-02       4.0
1167830  US3165001070         FDUS_NASD 2019-02-02     34 2019-02-05       3.0
816559   US1416331072         CRCM_NYSE 2019-02-12     11 2019-02-15       3.0
2163174  US86722A1034          SXC_NYSE 2019-01-26     34 2019-01-29       3.0
52014    CA1363851017       CNQ_Toronto 2019-03-05     35 2019-03-08       3.0
101110   CA87262K1057         X_Toronto 2019-02-25     39 2019-02-28       3.0

In [187]:
a = df_tmp.isna().sum().reset_index()
a.columns=['ticker','miss_count']

In [193]:
a = a[a['ticker']=='isin']

In [197]:
a = a.replace('isin','2002_Taiwan')

ticker  miss_count
0  2002_Taiwan           1

In [201]:
Date_gap_list=[]

for n in ticker_check_list:

    df_tmp = engage_new[engage_new['ticker']== n]
    df_tmp[['isin','score']] = df_tmp.groupby('ticker')['isin','score'].shift(-1)
    a = df_tmp.isna().sum().reset_index()
    a.columns=['ticker','miss_count']
    a = a[a['ticker']=='isin']
    Date_gap_list_tmp = a.replace('isin',n)
    
    Date_gap_list.append(Date_gap_list_tmp)
    
    
    

/Users/yangyufan/anaconda3/lib/python3.6/site-packages/pandas/core/frame.py:3140: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  self[k1] = value[k2]


In [206]:
Date_gap_list

[        ticker  miss_count
 0  2002_Taiwan           1,         ticker  miss_count
 0  2332_Taiwan           1,        ticker  miss_count
 0  3771_Tokyo           1,        ticker  miss_count
 0  3842_Tokyo           1,              ticker  miss_count
 0  4280_SaudiArabia           1,        ticker  miss_count
 0  6664_Tokyo           1,        ticker  miss_count
 0  6676_Tokyo           1,        ticker  miss_count
 0  8247_Tokyo           1,        ticker  miss_count
 0  8795_Tokyo           1,       ticker  miss_count
 0  ACXM_NASD           1,      ticker  miss_count
 0  AET_NYSE           1,        ticker  miss_count
 0  AGM.A_NYSE           1,      ticker  miss_count
 0  ANW_NYSE           1,               ticker  miss_count
 0  ARG_EuronextParis           1,      ticker  miss_count
 0  BD15_LSE           1,     ticker  miss_count
 0  BEZ_LSE           1,     ticker  miss_count
 0  BLT_LSE           1,       ticker  miss_count
 0  BOFI_NASD           1,        ticker  miss_count

In [59]:
count = engage.groupby(['symbol'])['date'].count().sort_values().reset_index()

In [60]:
fig = plt.figure(figsize=(20, 12))
sns.set(style = "ticks", color_codes = True)
plt.xticks(rotation = 'vertical')
ax = sns.lineplot(x="symbol", y="date", data=count, sort=False)